### ISSUE TITLE
How to slide a specified coordinate point?

### ISSUE THREAD

**Issue by GanZhiXiong** [entronad/graphic#292]
When there are many coordinate points, the icon supports horizontal scrolling by setting horizontalRange.

**So how do you slide to a specified coordinate point outside the visual area?**

**Comment by entronad** [comment]
Of course, I'd be happy to help you with that. As the creator of Graphic, I designed it to be flexible for handling large datasets and dynamic interactions like scrolling to a specific point. Here’s a detailed explanation of how you can achieve this.

When you have a chart with many data points and you've enabled horizontal scrolling by setting `horizontalRange`, the chart's visible area is essentially a "viewport" into your full dataset. To programmatically scroll to a specific data point, whether it's currently visible or not, you need to update this viewport.

The key to this functionality lies in manipulating the `horizontalRange` property of the `Coord` component in your `Chart`. The `horizontalRange` is a `List<double>` with two values, `[start, end]`, where both `start` and `end` are between `0.0` and `1.0`. These values represent the portion of the horizontal axis that is currently visible.

* `[0.0, 1.0]` means the entire dataset is visible.
* `[0.0, 0.5]` means the first half of the dataset is visible.
* `[0.5, 1.0]` means the second half of the dataset is visible.

To scroll to a specific data point, you need to calculate a new `horizontalRange` that centers the viewport on that point and then update the state of your chart.

***

### Principle

1.  **Identify the Target Index**: First, you need to know the index of the data point you want to scroll to. Let's call this `targetIndex`.
2.  **Determine the Viewport Size**: You need to decide how much of the data range you want to keep visible after scrolling. This is determined by the difference between the `end` and `start` of your `horizontalRange`. Let's call this `viewportFraction`. For example, if you want to show 20% of your data at a time, `viewportFraction` would be `0.2`.
3.  **Calculate the Center Position**: The position of your target data point within the entire dataset can be represented as a fraction. If you have `totalPoints`, the fractional position of your `targetIndex` is `targetIndex / (totalPoints - 1)`.
4.  **Calculate the New Range**: Now, you can calculate the new `start` and `end` for `horizontalRange`. The goal is to make the target point the center of the new viewport.
    * `newStart = targetPosition - (viewportFraction / 2)`
    * `newEnd = targetPosition + (viewportFraction / 2)`
5.  **Handle Edge Cases**: You must ensure that `newStart` is not less than `0.0` and `newEnd` is not greater than `1.0`.
    * If `newStart < 0.0`, set `newStart = 0.0` and `newEnd = viewportFraction`.
    * If `newEnd > 1.0`, set `newEnd = 1.0` and `newStart = 1.0 - viewportFraction`.
6.  **Update the State**: Finally, you update the state variable that holds your `horizontalRange` and Flutter's state management will rebuild the chart with the new visible range, effectively scrolling to your target point.

***

### Code Example

Let's put this into a practical Flutter example. Imagine you have a line chart with a large number of data points and buttons to scroll to the beginning, middle, and a specific index.

First, make sure you have the `graphic` package in your `pubspec.yaml`.

```dart
import 'package:flutter/material.dart';
import 'package:graphic/graphic.dart';

class ScrollableChartExample extends StatefulWidget {
  const ScrollableChartExample({super.key});

  @override
  State<ScrollableChartExample> createState() => _ScrollableChartExampleState();
}

class _ScrollableChartExampleState extends State<ScrollableChartExample> {
  // Your full dataset.
  final List<Map<String, dynamic>> data = List.generate(100, (i) => {
        'day': 'Day ${i + 1}',
        'value': 10 + i + (i % 5) * 5 - (i % 10) * 2.0,
      });

  // State variable for the horizontal range.
  // We'll initialize it to show the first 10% of the data.
  List<double> _horizontalRange = [0.0, 0.1];

  // The size of our viewport, e.g., showing 10% of the data.
  final double _viewportFraction = 0.1;

  void _scrollToIndex(int index) {
    final int totalPoints = data.length;
    if (totalPoints <= 1) return;

    // Calculate the fractional position of the target index.
    final double targetPosition = index / (totalPoints - 1);

    // Calculate the new start and end of the range.
    double newStart = targetPosition - (_viewportFraction / 2);
    double newEnd = targetPosition + (_viewportFraction / 2);

    // Clamp the values to stay within the [0, 1] bounds.
    if (newStart < 0.0) {
      newStart = 0.0;
      newEnd = _viewportFraction;
    }
    if (newEnd > 1.0) {
      newEnd = 1.0;
      newStart = 1.0 - _viewportFraction;
    }

    // Ensure the range is valid, especially for small datasets.
    newEnd = newEnd > 1.0 ? 1.0 : newEnd;
    newStart = newStart < 0.0 ? 0.0 : newStart;

    setState(() {
      _horizontalRange = [newStart, newEnd];
    });
  }


  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('Graphic Scroll To Point'),
      ),
      body: Column(
        children: [
          Container(
            margin: const EdgeInsets.all(10),
            height: 300,
            child: Chart(
              data: data,
              variables: {
                'day': Variable(
                  accessor: (Map map) => map['day'] as String,
                ),
                'value': Variable(
                  accessor: (Map map) => map['value'] as num,
                ),
              },
              elements: [
                LineElement(
                  shape: ShapeAttr(value: BasicLineShape(smooth: true)),
                )
              ],
              axes: [
                AxisGuide(
                  line: StrokeStyle(width: 1),
                ),
                AxisGuide(
                  line: StrokeStyle(width: 1),
                )
              ],
              coord: RectCoord(
                horizontalRange: _horizontalRange,
                // The gesture zoom/pan will update this range.
                // We use a listener to sync our state variable.
                horizontalRangeUpdater: (range) {
                  // Avoid unnecessary rebuilds if the range hasn't changed.
                  if (range[0] != _horizontalRange[0] || range[1] != _horizontalRange[1]) {
                    setState(() {
                      _horizontalRange = range;
                    });
                  }
                },
              ),
              interactions: [
                PointInteraction(
                  onTap: (point) {
                    print('Tapped on index: ${point.first.index}');
                  }
                ),
              ],
              gesture: Gesture(
                onHorizontalDragUpdate: (details) => true, // Enable panning
                onScaleUpdate: (details) => true,      // Enable zooming
              ),
            ),
          ),
          const SizedBox(height: 20),
          Wrap(
            spacing: 10,
            runSpacing: 10,
            alignment: WrapAlignment.center,
            children: [
              ElevatedButton(
                onPressed: () => _scrollToIndex(0),
                child: const Text('Scroll to Start (Index 0)'),
              ),
              ElevatedButton(
                onPressed: () => _scrollToIndex(49),
                child: const Text('Scroll to Middle (Index 49)'),
              ),
              ElevatedButton(
                onPressed: () => _scrollToIndex(99),
                child: const Text('Scroll to End (Index 99)'),
              ),
              ElevatedButton(
                onPressed: () => _scrollToIndex(75),
                child: const Text('Scroll to Index 75'),
              ),
            ],
          ),
        ],
      ),
    );
  }
}
```

### Explanation of the Code

1.  **`_horizontalRange` State**: We define a state variable `_horizontalRange` to hold the current visible range. This is passed directly to the `RectCoord`.
2.  **`_scrollToIndex(int index)` Method**: This is the core logic. It takes a target index, calculates the new range as described in the "Principle" section, and calls `setState` to apply the change.
3.  **`horizontalRangeUpdater`**: This is a crucial callback on `RectCoord`. When the user manually pans or zooms the chart, Graphic internally updates the range. This updater is called with the new range, allowing you to synchronize your `_horizontalRange` state variable with the chart's internal state. This ensures that your programmatic scrolling and user gestures work together seamlessly.
4.  **Buttons**: The `ElevatedButton` widgets simply call `_scrollToIndex` with predefined indices to demonstrate the functionality.

By following this approach, you can easily implement programmatic scrolling to any data point in your chart, providing a much more interactive and user-friendly experience when dealing with large datasets.

Feel free to check out the official documentation for more details on `Coord` and other components: [[Graphic Documentation](https://pub.dev/documentation/graphic/latest/graphic/)](https://pub.dev/documentation/graphic/latest/graphic/).

I hope this detailed explanation helps you effectively implement this feature in your Flutter application. Let me know if you have any more questions!
